### PR TITLE
Loosen SparseMatrixCSC colptr and rowval integer type restriction

### LINKED ISSUE #31435: SparseMatrixCSC type check is too restrictive (regression)
A little under a month ago #31118 was merged, in order to address #31024. It introduced checks in the `SparseMatrixCSC` constructor to ensure that the integer type used to store the `rowval` and `colptr` arrays was wide enough. The function that performs the check is https://github.com/JuliaLang/julia/blob/54fe55b370c23911f50e269ba59fa25e88cac642/stdlib/SparseArrays/src/sparsematrix.jl#L36

The function has the line
```
!isbitstype(Ti) || n*m+1 ≤ typemax(Ti) || throwTi("maximal nnz+1", "m*n+1", n*m+1)
```
which checks that `Ti` is wide enough to store the largest possible sparse matrix of size m X n, aka a dense matrix of size m X n. Shouldn't that line be
```
!isbitstype(Ti) || nnz+1 ≤ typemax(Ti) || throwTi("maximal nnz+1", "m*n+1", n*m+1)
```
where `nnz = length(nzval)` is the number of nonzeros entries in the matrix? Is there somewhere else in the `SparseArrays` stdlib where the `rowval` and `colptr` arrays need to grow for some reason? Aside from that I can't think of why the restriction above was implemented. It has broken some of my company's code where we have some quite large and very sparse matrices where we store the `rowval` and `colptr` arrays using a small integer type to save memory. The following MWE works on julia 1.1 but fails on master built from source yesterday (Commit f611b46e72):
```
julia> using SparseArrays, LinearAlgebra
julia> colptr = fill(2,101);
julia> colptr[1] = 1;
julia> rowval = [1];
A = SparseMatrixCSC(100,100,Int8.(colptr),Int8.(rowval),[2.2])
ERROR: ArgumentError: maximal nnz+1 (m*n+1 = 10001) does not fit in Ti = Int8
Stacktrace:
 [1] (::getfield(SparseArrays, Symbol("#throwTi#2")){DataType})(::String, ::String, ::Int64) at /home/patrick/Julia/master/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:39
 [2] sparse_check_Ti at /home/patrick/Julia/master/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:45 [inlined]
 [3] Type at /home/patrick/Julia/master/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:26 [inlined]
 [4] SparseMatrixCSC(::Int64, ::Int64, ::Array{Int8,1}, ::Array{Int8,1}, ::Array{Float64,1}) at /home/patrick/Julia/master/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:33
 [5] top-level scope at REPL[13]:1
```

Also note that the `sparse_check_Ti` function as written is unsafe because m and n can be any integer types and `m*n` will return a spurious result if m and n are of an integer type that can't hold the result of `m*n`.

I can submit a PR to change this unless there's some reason I'm not thinking of that the current behaviour, or at least something more restrictive than `nnz+1 ≤ typemax(Ti) ` is required.

### PR DESCRIPTION
**Pull request by Pbellive**
This fixes #31435. Please see that issue for full explanation. @KlausC any objection to this change?

### CODE REVIEW

**Review by mbauman — APPROVED** [review]

**Comment by ViralBShah** [comment]
I noted a potential objection in #31435.

**Review by KlausC — COMMENTED** [review]
should be: `nnzval < typemax(Ti)` for correct handling of corner case `Ti = Int; nnzval = typemax(Ti)`.

error text `"Number .."` should be lowercase for consistency with other messages.

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 1]
```diff
@@ -42,7 +42,7 @@ function sparse_check_Ti(m::Integer, n::Integer, Ti::Type)
         n < 0 && throwsz("columns", 'n', n)
         !isbitstype(Ti) || m ≤ typemax(Ti) || throwTi("number of rows", "m", m)
         !isbitstype(Ti) || n ≤ typemax(Ti) || throwTi("number of columns", "n", n)
-        !isbitstype(Ti) || n*m+1 ≤ typemax(Ti) || throwTi("maximal nnz+1", "m*n+1", n*m+1)
+        !isbitstype(Ti) || nnzval+1 ≤ typemax(Ti) || throwTi("Number of nonzeros", "nnz+1", nnzval+1)
```
Comment (fredrikekre) [inline stdlib/SparseArrays/src/sparsematrix.jl]:
```suggestion
        !isbitstype(Ti) || nnzval < typemax(Ti) || throwTi("number of nonzeros", "nnz+1", nnzval+1)
```

**Comment by KlausC** [comment]
As `nnz(A)` may change during the lifetime of `A`, it would be wise to assert that the value we want to store in `A.colptr[A.n+1]`  (that is the correct new `nnz(A)+1`) fits in a `Ti`.

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 2]
```diff
@@ -23,7 +23,7 @@ struct SparseMatrixCSC{Tv,Ti<:Integer} <: AbstractSparseMatrix{Tv,Ti}
     function SparseMatrixCSC{Tv,Ti}(m::Integer, n::Integer, colptr::Vector{Ti}, rowval::Vector{Ti},
                                     nzval::Vector{Tv}) where {Tv,Ti<:Integer}
 
-        sparse_check_Ti(m, n, Ti)
+        sparse_check_Ti(m, n, length(nzval), Ti)
```
Comment (KlausC) [inline stdlib/SparseArrays/src/sparsematrix.jl]:
```suggestion
        sparse_check_Ti(m, n, max(length(nzval), length(rowval)), Ti)
```

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 3]
```diff
@@ -594,7 +594,7 @@ function sparse!(I::AbstractVector{Ti}, J::AbstractVector{Ti},
         csccolptr::Vector{Ti}, cscrowval::Vector{Ti}, cscnzval::Vector{Tv}) where {Tv,Ti<:Integer}
 
     require_one_based_indexing(I, J, V)
-    sparse_check_Ti(m, n, Ti)
+    sparse_check_Ti(m, n, length(V), Ti)
```
Comment (KlausC) [inline stdlib/SparseArrays/src/sparsematrix.jl]:
```suggestion
    sparse_check_Ti(m, n, length(I), Ti)
```

**Comment by ViralBShah** [comment]
Can't `setindex!` potentially grow the size of `nzval` and `rowval`, thus requiring the check to be the way it is right now (but being careful about the overflow)?

**Comment by mbauman** [comment]
Yes, but:

* That's not how things behaved prior to #31118
* The change in #31118 was breaking
* We merged #31118 without consideration of the breaking-ness

So these sorts of "it'd be nice to guarantee"s are non-sequitors.  We just need to get things back to the old status quo and then we can discuss how to approach the next steps.

It's either merge this patch or wholesale revert #31118.

**Comment by KristofferC** [comment]
The discussion here shows that there are intricacies we don't really have time to deal with now, considering we want to branch for 1.3 about now. I suggest just reverting the breaking PR (https://github.com/JuliaLang/julia/pull/31667) and then the discussion can take as long as it wants to add it back in a non-breaking way.

**Comment by KlausC** [comment]
Sorry, but without a replacement for #31118 the software is breaking because of #31024. Propose to merge this PR with the proposed changes.

**Comment by KristofferC** [comment]
> the software is breaking because of

What code would reverting the PR break? I see an error being replaced with another one.

**Comment by mbauman** [comment]
Note that we're using breaking with a very specific meaning here: it's not the Julia is "broken" (e.g., with a seg fault in some instances) but rather that a particular change caused a regression that broke someone's workflow that previously worked.  That's significantly worse than an existing bug that has been a bug for a long time.

**Comment by Pbellive** [comment]
I'm in favour of @KristofferC 's suggestion to revert #31118 and continue discussion of a long term fix here. @ViralBShah and @KlausC 's comments that `setindex!` can grow the internal arrays are of course important points. To address that, it seems like it would be sufficient to just add a type check inside the setindex! methods to make sure that `A.colptr[n+1] < typemax(Ti)` remains true after the new entries are added to the matrix. I haven't tested that but I wouldn't think it would be much of a performance hit.

**Comment by mauro3** [comment]
> Sorry, but without a replacement for #31118 the software is breaking because of #31024

@KlausC: until it is fixed you can monkey-patch it: https://discourse.julialang.org/t/difficulties-with-recent-julia-releases/22969/2)

**Comment by KlausC** [comment]
@mbauman 
> Note that we're using breaking with a very specific meaning here: ... a particular change caused a regression that broke someone's workflow that previously worked

Does that mean, we don't introduce data consistency checks, because someone's program worked, although the data, it operated on, were partially corrupted?

In our case, I agree, that the introduced check was too restrictive, but generally ...?

**Comment by mbauman** [comment]
It just means that we need to be careful about doing so.  We flag all such changes as [minor changes](https://github.com/JuliaLang/julia/issues?q=label%3A"minor+change"), we ensure they appear in NEWS.md, and such changes have a higher bar to cross to get merged.

In cases where the previous behavior resulted in corrupted or buggy answers, then that's not breaking — it's a bug.  Admittedly sometimes the line between the two can be blurry.

In this case, I still really want to see a resurrected #31118+#31661 that fixes the segfault and protects against overfilling such arrays.  Had we not been right on a branching threshold, we probably would have let things percolate a little longer without a direct revert.

**Comment by ViralBShah** [comment]
In general, we always want correctness, but this was a tricky situation as already discussed above. Reverting and redoing is certainly the right thing.

**Comment by Pbellive** [comment]
I'm happy to update and expand this PR to make an attempt at addressing the concerns raised in this thread. I should be able to get to it in the next 4-5 days.

**Comment by KlausC** [comment]
@Pbellive sorry for taking your job. I hope you can live with this solution.

**Pull request closed without merging**